File: java/src/main/java/com/squirrelparser/Combinators.java
package com.squirrelparser;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static com.squirrelparser.MatchResult.mismatch;

/**
 * Helper: check if all children are complete.
 */
final class CombinatorHelper {
    static boolean allComplete(List<MatchResult> children) {
        return children.stream().allMatch(c -> c.isMismatch() || c.isComplete());
    }
}

// -----------------------------------------------------------------------------------------------------------------

/**
 * Sequence: matches all sub-clauses in order, with error recovery.
 */
final class Seq extends HasMultipleSubClauses {
    public Seq(List<Clause> subClauses) {
        super(subClauses);
    }

    @Override
    public MatchResult match(Parser parser, int pos, Clause bound) {
        List<MatchResult> children = new ArrayList<>();
        int curr = pos;
        int i = 0;

        while (i < subClauses.size()) {
            Clause clause = subClauses.get(i);
            Clause next = (i + 1 < subClauses.size()) ? subClauses.get(i + 1) : null;
            Clause effectiveBound = (parser.inRecoveryPhase() && next != null) ? next : bound;
            MatchResult result = parser.match(clause, curr, effectiveBound);

            if (result.isMismatch()) {
                if (parser.inRecoveryPhase() && !result.isFromLRContext()) {
                    var recovery = recover(parser, curr, i);
                    if (recovery != null) {
                        ParserStats.recordRecovery();
                        int inputSkip = recovery.inputSkip;
                        int grammarSkip = recovery.grammarSkip;
                        MatchResult probe = recovery.probe;

                        if (inputSkip > 0) {
                            children.add(new SyntaxError(curr, inputSkip));
                        }

                        for (int j = 0; j < grammarSkip; j++) {
                            children.add(new SyntaxError(curr + inputSkip, 0, subClauses.get(i + j)));
                        }

                        if (probe == null) {
                            curr += inputSkip;
                            break;
                        }

                        children.add(probe);
                        curr += inputSkip + probe.len();
                        i += grammarSkip + 1;
                        continue;
                    }
                }
                return mismatch();
            }

            children.add(result);
            curr += result.len();
            i++;
        }

        if (children.isEmpty()) {
            return new Match(this, pos, 0);
        }

        return Match.withChildren(this, children, CombinatorHelper.allComplete(children));
    }

    private record Recovery(int inputSkip, int grammarSkip, MatchResult probe) {}

    private Recovery recover(Parser parser, int curr, int i) {
        int maxScan = parser.input().length() - curr + 1;
        int maxGrammar = subClauses.size() - i;

        for (int inputSkip = 0; inputSkip < maxScan; inputSkip++) {
            int probePos = curr + inputSkip;

            if (probePos >= parser.input().length()) {
                if (inputSkip == 0) {
                    return new Recovery(inputSkip, maxGrammar, null);
                }
                continue;
            }

            for (int grammarSkip = 0; grammarSkip < maxGrammar; grammarSkip++) {
                if (grammarSkip == 0 && inputSkip == 0) continue;
                if (grammarSkip > 0) continue;

                int clauseIdx = i + grammarSkip;
                Clause clause = subClauses.get(clauseIdx);

                Clause failedClause = subClauses.get(i);
                if (failedClause instanceof Str str && str.text().length() == 1 && inputSkip > 1) {
                    if (clauseIdx + 1 < subClauses.size()) {
                        Clause nextClause = subClauses.get(clauseIdx + 1);
                        if (nextClause instanceof Str nextStr) {
                            String skipped = parser.input().substring(curr, curr + inputSkip);
                            if (skipped.contains(nextStr.text())) {
                                continue;
                            }
                        }
                    }
                }
                MatchResult probe = parser.probe(clause, probePos);
                if (!probe.isMismatch()) {
                    if (clause instanceof Str str && inputSkip > str.text().length()) {
                        if (str.text().length() > 1) {
                            continue;
                        }
                        String skipped = parser.input().substring(curr, curr + inputSkip);
                        if (skipped.contains(str.text())) {
                            continue;
                        }
                    }
                    return new Recovery(inputSkip, grammarSkip, probe);
                }
            }
        }
        return null;
    }

    @Override
    public String toString() {
        return "(" + subClauses.stream().map(Object::toString).collect(Collectors.joining(" ")) + ")";
    }
}

// -----------------------------------------------------------------------------------------------------------------

/**
 * Ordered choice: matches the first successful sub-clause.
 */
final class First extends HasMultipleSubClauses {
    public First(List<Clause> subClauses) {
        super(subClauses);
    }

    @Override
    public MatchResult match(Parser parser, int pos, Clause bound) {
        for (int i = 0; i < subClauses.size(); i++) {
            MatchResult result = parser.match(subClauses.get(i), pos, bound);
            if (!result.isMismatch()) {
                if (parser.inRecoveryPhase() && i == 0 && result.totDescendantErrors() > 0) {
                    MatchResult bestResult = result;
                    int bestLen = result.len();
                    int bestErrors = result.totDescendantErrors();

                    for (int j = 1; j < subClauses.size(); j++) {
                        MatchResult altResult = parser.match(subClauses.get(j), pos, bound);
                        if (!altResult.isMismatch()) {
                            int altLen = altResult.len();
                            int altErrors = altResult.totDescendantErrors();

                            double bestErrorRate = bestLen > 0 ? (double) bestErrors / bestLen : 0.0;
                            double altErrorRate = altLen > 0 ? (double) altErrors / altLen : 0.0;
                            double errorRateThreshold = 0.5;

                            if ((bestErrorRate >= errorRateThreshold && altErrorRate < errorRateThreshold) ||
                                altLen > bestLen ||
                                (altLen == bestLen && altErrors < bestErrors)) {
                                bestResult = altResult;
                                bestLen = altLen;
                                bestErrors = altErrors;
                            }
                            if (altErrors == 0 && altLen >= bestLen) break;
                        }
                    }
                    return Match.withChildren(this, List.of(bestResult), bestResult.isComplete());
                }
                return Match.withChildren(this, List.of(result), result.isComplete());
            }
        }
        return mismatch();
    }

    @Override
    public String toString() {
        return "(" + subClauses.stream().map(Object::toString).collect(Collectors.joining(" / ")) + ")";
    }
}

// -----------------------------------------------------------------------------------------------------------------

/**
 * Base class for repetition (OneOrMore, ZeroOrMore).
 */
sealed class Repetition extends HasOneSubClause permits OneOrMore, ZeroOrMore {
    private final boolean requireOne;

    protected Repetition(Clause subClause, boolean requireOne) {
        super(subClause);
        this.requireOne = requireOne;
    }

    @Override
    public MatchResult match(Parser parser, int pos, Clause bound) {
        List<MatchResult> children = new ArrayList<>();
        int curr = pos;
        boolean incomplete = false;
        boolean hasRecovered = false;

        while (curr <= parser.input().length()) {
            if (parser.inRecoveryPhase() && bound != null) {
                if (parser.canMatchNonzeroAt(bound, curr)) {
                    break;
                }
            }

            MatchResult result = parser.match(subClause, curr);
            if (result.isMismatch()) {
                if (!parser.inRecoveryPhase() && curr < parser.input().length()) {
                    incomplete = true;
                }

                if (parser.inRecoveryPhase()) {
                    var recovery = recover(parser, curr, hasRecovered);
                    if (recovery != null) {
                        ParserStats.recordRecovery();
                        int skip = recovery.skip;
                        MatchResult probe = recovery.probe;
                        children.add(new SyntaxError(curr, skip));
                        hasRecovered = true;
                        if (probe != null) {
                            children.add(probe);
                            curr += skip + probe.len();
                            continue;
                        } else {
                            curr += skip;
                            break;
                        }
                    }
                }
                break;
            }
            if (result.len() == 0) break;
            children.add(result);
            curr += result.len();
        }
        if (requireOne && children.isEmpty()) {
            return mismatch();
        }
        if (children.isEmpty()) {
            return new Match(this, pos, 0, List.of(), !incomplete, false, 0);
        }
        return Match.withChildren(this, children, !incomplete && CombinatorHelper.allComplete(children));
    }

    private record RepetitionRecovery(int skip, MatchResult probe) {}

    private RepetitionRecovery recover(Parser parser, int curr, boolean hasRecovered) {
        for (int skip = 1; skip < parser.input().length() - curr + 1; skip++) {
            MatchResult probe = parser.probe(subClause, curr + skip);
            if (!probe.isMismatch()) {
                return new RepetitionRecovery(skip, probe);
            }
        }
        if (hasRecovered && curr < parser.input().length()) {
            int skipToEnd = parser.input().length() - curr;
            return new RepetitionRecovery(skipToEnd, null);
        }
        return null;
    }
}

// -----------------------------------------------------------------------------------------------------------------

/**
 * One or more repetitions.
 */
final class OneOrMore extends Repetition {
    public OneOrMore(Clause subClause) {
        super(subClause, true);
    }

    @Override
    public String toString() {
        return subClause + "+";
    }
}

// -----------------------------------------------------------------------------------------------------------------

/**
 * Zero or more repetitions.
 */
final class ZeroOrMore extends Repetition {
    public ZeroOrMore(Clause subClause) {
        super(subClause, false);
    }

    @Override
    public String toString() {
        return subClause + "*";
    }
}

// -----------------------------------------------------------------------------------------------------------------

/**
 * Optional: matches zero or one instance.
 */
final class Optional extends HasOneSubClause {
    public Optional(Clause subClause) {
        super(subClause);
    }

    @Override
    public MatchResult match(Parser parser, int pos, Clause bound) {
        MatchResult result = parser.match(subClause, pos, bound);

        if (result.isMismatch()) {
            boolean incomplete = !parser.inRecoveryPhase() && pos < parser.input().length();
            return new Match(this, pos, 0, List.of(), !incomplete, false, 0);
        }

        return Match.withChildren(this, List.of(result), result.isComplete());
    }

    @Override
    public String toString() {
        return subClause + "?";
    }
}

// -----------------------------------------------------------------------------------------------------------------

/**
 * Reference to a named rule.
 */
final class Ref extends Clause {
    private final String ruleName;

    public Ref(String ruleName) {
        this.ruleName = ruleName;
    }

    public String ruleName() {
        return ruleName;
    }

    @Override
    public MatchResult match(Parser parser, int pos, Clause bound) {
        Clause clause = parser.rules().get(ruleName);
        if (clause == null) {
            throw new IllegalArgumentException("Rule \"" + ruleName + "\" not found");
        }
        MatchResult result = parser.match(clause, pos, bound);
        if (result.isMismatch()) return result;
        return Match.withChildren(this, List.of(result), result.isComplete());
    }

    @Override
    public void checkRuleRefs(Map<String, Clause> grammarMap) {
        if (!grammarMap.containsKey(ruleName) && !grammarMap.containsKey("~" + ruleName)) {
            throw new IllegalArgumentException("Rule \"" + ruleName + "\" not found in grammar");
        }
    }

    @Override
    public String toString() {
        return ruleName;
    }
}

// -----------------------------------------------------------------------------------------------------------------

/**
 * Negative lookahead: succeeds if sub-clause fails, consumes nothing.
 */
final class NotFollowedBy extends HasOneSubClause {
    public NotFollowedBy(Clause subClause) {
        super(subClause);
    }

    @Override
    public MatchResult match(Parser parser, int pos, Clause bound) {
        MatchResult result = parser.match(subClause, pos, bound);
        return result.isMismatch() ? new Match(this, pos, 0) : mismatch();
    }

    @Override
    public String toString() {
        return "!" + subClause;
    }
}

// -----------------------------------------------------------------------------------------------------------------

/**
 * Positive lookahead: succeeds if sub-clause succeeds, consumes nothing.
 */
final class FollowedBy extends HasOneSubClause {
    public FollowedBy(Clause subClause) {
        super(subClause);
    }

    @Override
    public MatchResult match(Parser parser, int pos, Clause bound) {
        MatchResult result = parser.match(subClause, pos, bound);
        return result.isMismatch() ? mismatch() : new Match(this, pos, 0);
    }

    @Override
    public String toString() {
        return "&" + subClause;
    }
}

File: dart/lib/src/parser/combinators.dart
import 'clause.dart';
import 'match_result.dart';
import 'parser.dart';

// -----------------------------------------------------------------------------------------------------------------

abstract class HasOneSubClause extends Clause {
  final Clause subClause;
  const HasOneSubClause(this.subClause);

  @override
  void checkRuleRefs(Map<String, Clause> grammarMap) {
    subClause.checkRuleRefs(grammarMap);
  }
}

abstract class HasMultipleSubClauses extends Clause {
  final List<Clause> subClauses;
  const HasMultipleSubClauses(this.subClauses);

  @override
  void checkRuleRefs(Map<String, Clause> grammarMap) {
    for (final clause in subClauses) {
      clause.checkRuleRefs(grammarMap);
    }
  }
}

// -----------------------------------------------------------------------------------------------------------------

/// Sequence: matches all sub-clauses in order.
class Seq extends HasMultipleSubClauses {
  const Seq(super.subClauses);

  @override
  MatchResult match(Parser parser, int pos) {
    final children = <MatchResult>[];
    int curr = pos;

    for (int i = 0; i < subClauses.length; i++) {
      final result = subClauses[i].match(parser, curr);
      if (result.isMismatch) {
        // THE SLOTS THAT DID MATCH ARE THE FRONTIER, and the one that did not
        // is where recovery has to look. Keep both: the matched prefix as the
        // input this sequence accounted for, and the failing slot's own
        // mismatch as the child to descend into. PEG stops at the first failing
        // slot, so there is exactly one of those, and it is always the last
        // child -- which is why it can be appended rather than copied into a
        // second list.
        children.add(result);
        return Mismatch(this, pos, curr - pos, subClauseMatches: children);
      }
      children.add(result);
      curr += result.len;
    }

    if (children.isEmpty) {
      return Match(this, pos, 0);
    }
    return Match(this, 0, 0, subClauseMatches: children);
  }

  @override
  String toString() => '(${subClauses.join(' ')})';
}

// -----------------------------------------------------------------------------------------------------------------

/// Ordered choice: matches the first successful sub-clause.
class First extends HasMultipleSubClauses {
  const First(super.subClauses);

  @override
  MatchResult match(Parser parser, int pos) {
    // Allocated only once an arm has failed, so an ordered choice that takes
    // its first arm allocates nothing it did not allocate before. (Measured on
    // its own this was inside run noise; the cost that did show up was the
    // per-memo-hit wrapper in [Ref], below.)
    List<MatchResult>? failed;
    var read = 0;
    for (int i = 0; i < subClauses.length; i++) {
      final result = subClauses[i].match(parser, pos);
      if (!result.isMismatch) {
        return Match(this, 0, 0, subClauseMatches: [result]);
      }
      // EVERY ARM IS KEPT, because a choice whose arms all failed is not itself
      // a place to repair -- the repair belongs in whichever arm got furthest,
      // and which that is cannot be known from here. So this reports the
      // furthest any arm read and hands all of them on to be descended into.
      (failed ??= <MatchResult>[]).add(result);
      if (result.len > read) read = result.len;
    }
    return Mismatch(this, pos, read, subClauseMatches: failed ?? const []);
  }

  @override
  String toString() => '(${subClauses.join(' / ')})';
}

// -----------------------------------------------------------------------------------------------------------------

/// Base class for repetition (OneOrMore, ZeroOrMore).
class Repetition extends HasOneSubClause {
  final bool requireOne;

  const Repetition(super.subClause, {required this.requireOne});

  @override
  MatchResult match(Parser parser, int pos) {
    final children = <MatchResult>[];
    int curr = pos;
    MatchResult? stoppedBy;

    while (curr <= parser.input.length) {
      final result = subClause.match(parser, curr);
      if (result.isMismatch) {
        stoppedBy = result;
        break;
      }
      // Never consume more than one zero-length subclause match, to prevent
      // an infinite loop (e.g. for the pathological grammar "()*").
      if (result.len == 0) break;
      children.add(result);
      curr += result.len;
    }

    if (requireOne && children.isEmpty) {
      // Nothing repeated even once, so nothing was read; the body's own
      // mismatch is the whole account of why.
      return Mismatch(this, pos, 0, subClauseMatches: stoppedBy == null ? const [] : [stoppedBy]);
    }
    if (children.isEmpty) {
      return Match(this, pos, 0);
    }
    return Match(this, 0, 0, subClauseMatches: children);
  }
}

/// One or more repetitions.
class OneOrMore extends Repetition {
  const OneOrMore(super.subClause) : super(requireOne: true);

  @override
  String toString() => '$subClause+';
}

/// Zero or more repetitions.
class ZeroOrMore extends Repetition {
  const ZeroOrMore(super.subClause) : super(requireOne: false);

  @override
  String toString() => '$subClause*';
}

// -----------------------------------------------------------------------------------------------------------------

/// Optional: matches zero or one instance.
class Optional extends HasOneSubClause {
  const Optional(super.subClause);

  @override
  MatchResult match(Parser parser, int pos) {
    final result = subClause.match(parser, pos);
    if (result.isMismatch) {
      return Match(this, pos, 0);
    }
    return Match(this, 0, 0, subClauseMatches: [result]);
  }

  @override
  String toString() => '$subClause?';
}

// -----------------------------------------------------------------------------------------------------------------

/// Reference to a named rule.
///
/// Ref is the only clause type that recurses back through the memo table (by
/// calling [Parser.match]), so memoization is triggered only at the level of
/// rules, rather than at the level of individual clauses.
class Ref extends Clause {
  final String ruleName;
  const Ref(this.ruleName);

  @override
  MatchResult match(Parser parser, int pos) {
    // The Parser strips the '~' prefix from rule names, so we can look up directly
    final clause = parser.rules[ruleName];
    if (clause == null) {
      throw ArgumentError('Rule "$ruleName" not found');
    }
    final result = parser.match(clause, pos);
    // A FAILED REFERENCE IS NOT A SECOND FAILURE, so it is passed through
    // rather than wrapped. The rule's own mismatch already has this position,
    // this consumed length, and -- in [clause] -- the memo key that says which
    // cell to re-attempt, which is what a frontier walk actually needs; a `Ref`
    // node on top of it would only repeat the rule name. Wrapping cost 15% of
    // parse time (measured), because it allocated a node and a list on every
    // memo HIT for a failing rule, and a hit is meant to be free. The success
    // path still wraps: [buildAST] reads the rule name off that node.
    if (result.isMismatch) return result;
    return Match(this, 0, 0, subClauseMatches: [result]);
  }

  @override
  void checkRuleRefs(Map<String, Clause> grammarMap) {
    if (!grammarMap.containsKey(ruleName) && !grammarMap.containsKey('~$ruleName')) {
      throw FormatException('Rule "$ruleName" not found in grammar');
    }
  }

  @override
  String toString() => ruleName;
}

// -----------------------------------------------------------------------------------------------------------------

/// Negative lookahead: succeeds if sub-clause fails, consumes nothing.
class NotFollowedBy extends HasOneSubClause {
  const NotFollowedBy(super.subClause);

  @override
  MatchResult match(Parser parser, int pos) {
    final result = subClause.match(parser, pos);
    // A PREDICATE READS NOTHING, SO ITS FRONTIER IS ITS OWN POSITION. This one
    // fails because its body SUCCEEDED, and that success is not input this
    // clause accepted -- it is the reason for the failure. Reporting the body's
    // length here would claim a frontier past input the enclosing sequence
    // never consumed. The body is not carried either: a repair placed inside a
    // zero-width assertion would consume input the assertion does not.
    return result.isMismatch ? Match(this, pos, 0) : Mismatch(this, pos, 0);
  }

  @override
  String toString() => '!$subClause';
}

// -----------------------------------------------------------------------------------------------------------------

/// Positive lookahead: succeeds if sub-clause succeeds, consumes nothing.
class FollowedBy extends HasOneSubClause {
  const FollowedBy(super.subClause);

  @override
  MatchResult match(Parser parser, int pos) {
    final result = subClause.match(parser, pos);
    // Zero-width for the same reason as [NotFollowedBy], and the body is
    // withheld for the same reason -- even though here the body genuinely did
    // fail and looks like a frontier. It is not one: satisfying `&Foo` needs
    // input that this clause would then not consume, so a syntax error span
    // placed under it would be spanned by a node of width 0.
    return result.isMismatch ? Mismatch(this, pos, 0) : Match(this, pos, 0);
  }

  @override
  String toString() => '&$subClause';
}
